src/hotspot/cpu/aarch64/register_aarch64.hpp

```diff
@@ -32,43 +32,64 @@
 class VMRegImpl;
 typedef VMRegImpl* VMReg;
 
+// Macros to help define all kinds of registers
+
+#define REGISTER_IMPL_DECLARATION(type, name)                           \
+inline const type as_ ## type(int encoding) {                           \
+  assert(encoding <= name::number_of_declared_registers, "invalid register"); \
+  return encoding == -1 ? name::invalid() : name::first() + encoding;   \
+}                                                                       \
+extern name all_ ## type ## s[name::number_of_declared_registers] INTERNAL_VISIBILITY; \
+constexpr type first_ ## type = all_ ## type ## s;                      \
+inline constexpr type name::first() { return all_ ## type ## s; }
+
+#define REGISTER_IMPL_DEFINITION(type, name)                    \
+name all_ ## type ## s[name::number_of_declared_registers];
+
+#undef CONSTANT_REGISTER_DECLARATION
+#define CONSTANT_REGISTER_DECLARATION(type, name, value)        \
+constexpr type name = first_##type + value;
+
+
 // Use Register as shortcut
 class RegisterImpl;
 typedef RegisterImpl* Register;
 
-inline const Register as_Register(int encoding) {
-  return (Register)(intptr_t) encoding;
-}
+const Register as_Register(int encoding);
 
 class RegisterImpl: public AbstractRegisterImpl {
- public:
+  const friend Register as_Register(int encoding);
+
+private:
+  static constexpr Register first();
+  static constexpr Register invalid() { return first() + number_of_declared_registers; }
+
+public:
   enum {
-    number_of_registers         =   32,
-    number_of_byte_registers      = 32,
-    number_of_registers_for_jvmci = 34,  // Including SP and ZR.
-    max_slots_per_register = 2
+    number_of_registers       = 32,
+    number_of_byte_registers  = 32,
+    number_of_declared_registers = 34,  // Including SP and ZR.
+    max_slots_per_register    = 2
   };
 
   // derived registers, offsets, and addresses
-  Register successor() const                          { return as_Register(encoding() + 1); }
-
-  // construction
-  inline friend const Register as_Register(int encoding);
+  Register successor() const     { return as_Register(encoding() + 1); }
```

Comment:
Can we simply return `this + 1` here?
Thank you very much.

Reply:
Oh yeah, good point.